doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,192 @@
+console.  When you run a GUI application from your console shell, it is **not** allocated a console. The shell will
+**not** wait for it to exit before returning you to a prompt.
+
+There is a large class of applications that has diverse console allocation needs. Take Python (or perhaps Perl, Ruby, or
+Lua, or even our own VBScript). People author GUI applications in Python and Perl and VBScript, for better or for worse.
+
+Because they are console subsystem applications, any user double-clicking a shortcut to a Python or Perl application
+will be presented with a useless black box that the language runtime may choose to garbage collect, or may choose not
+to.
+
+Any user running that GUI application from a console shell will see their shell hang until the application terminates.
+
+All of these scripting languages worked around this by shipping two binaries each, identical in every way expect in
+their subsystem bits. python/pythonw, perl/perlw, ruby/rubyw, wscript/cscript.
+
+Likewise, PowerShell\[1\] is waiting to deal with this problem because they don't necessarily want to ship a `pwshw.exe`
+for all of their GUI-only authors.
+
+On the other side, you have mostly-GUI applications that want to print output to a console **if there is one
+connected**.  Sometimes they'll allocate their own (and therefore a new window) to display in, and sometimes they'll
+reattach to the one they could have inherited. VSCode does the latter, and so when you run `code` from CMD, and then
+`exit` CMD, your console window sticks around because VSCode is still attached to it. It will never print anything, so
+you just have a dead console window.
+
+There's a risk in reattaching, though. Given that the shell decides whether to wait based on the subsystem field, GUI
+subsystem applications that reattach to their owning consoles *just to print some text* end up stomping on the output of
+any shell that doesn't wait for them:
+
+```
+C:\> wt --help
+
+wt - the Windows Terminal
+C:\> Usage: [OPTIONS] ...
+```
+
+> _(the prompt is interleaved with the output)_
+
+## Solution Design
+
+I propose that we introduce a fusion manifest field, **consoleAllocationPolicy**, with the following values:
+
+* `inheritOnly`
+* `always`
+
+It would look (roughly) like this:
+
+```xml
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
+  <application>
+    <windowsSettings>
+      <consoleAllocationPolicy xmlns="http://schemas.microsoft.com/SMI/20XX/WindowsSettings">inheritOnly</consoleAllocationPolicy>
+    </windowsSettings>
+  </application>
+</assembly>
+```
+
+This field will apply consistent behavior across different image subsystems, with differing end user results:
+
+|               | `SUBSYSTEM_GUI`                                                       | `SUBSYSTEM_GUI`               |

```

Comment:
I suspect there's a typo here, as both columns being `GUI` would be challenging to understand.

Reply:
Indeed. I've pushed an update.